### DISCUSSION TITLE
Docker / memory / mail authentication /  ldap

### DISCUSSION THREAD

**Discussion by bpatath** [simple-login/app#736] [Self-hosting question]
Hello,

First, thank you so much for this amazing app ! 
I am self-hosting it on a virtual private server for my own mails and I love it !

I have several questions and/or feature request. I might be able to make some PR if you point me towards the right file/function.

1. Is there a reason you use three successive docker run (database migrations, app init, and then app) instead of handling it all in one run ? Usually, docker images contain an `entrypoint.sh` which is responsible of verifying the state of the app and do necessary command to both initialize it and then run it.

2. What is the purpose of the `sl` and `upload` volume ? Since this is an email service, I am not sure I understanding what will be stored, and thus, which type of storage/backup I should use. Maybe the answer to this question should be added to the README.

3. When I start both the agent and the web server, they use respectively around 190MB and 280 MB of memory, which seems really big. In fact, of all the apps and containers that I host, it is the biggest. Have you ever tried profiling the app to see where the memory is used and if there is a way to reduce it ? It can be quite important when self-hosting it on a server with limited resources such as a cheap VPS.

4. About memory, and this should probably be an issue instead, I noticed that the memory usage increases over time.
There is a [SO question](https://stackoverflow.com/questions/49991234/flask-app-memory-leak-caused-by-each-api-call) about memory leaks in Flask app. As I don't know anything about flask, do you think that SimpleLogin could have such memory leaks ?

5. It would be nice if SMTP authentication was supported. Right now, I am forced to include the SimpleLogin IP address in postfix's "mynetworks", but it is not always adapted. Another way would be to provide SimpleLogin with an SMTP username and password instead.

6. As I use an LDAP directory to manage my users, I would have liked if SimpleLogin was able to authenticate users using LDAP. Do you use a modular authentication library in which LDAP support could be added easily ?

**Comment by nguyenkims** [comment]
Hi, glad that you like SimpleLogin :). Here are the answers:

1) The reason for running the migration separately is if it fails, it wouldn't affect the currently running containers. For example, if the migration fails (it shouldn't), the app container should still run with the previous image. For the init container, this one only needs to be run once as it can take time.

2) The `sl/upload` is to store quarantine emails, i.e. emails that are bounced by your mailbox. Without it, you can't view a refused email. This is mentioned in the README 
<img width="392" alt="image" src="https://user-images.githubusercontent.com/3167780/148035315-7dd3b43b-d44b-464f-8ab6-96ef7cff1f4f.png">


3) Python isn't good at memory optimisation and its VM takes a lot of memory by default ... SimpleLogin doesn't store anything especially big or uses an excessive caching. You can tweak the app to spend less memory by running `gunicorn wsgi:app -b 0.0.0.0:7777 -w 1` (note the `-w 1` instead of `-w 2`) which reduces the number of gunicorn workers and reduce the memory by half. It's recommended to have at least 2 web workers though for high availability.

4) We weren't aware of this issue and haven't noticed any memory leak on our servers but if you want to make sure, you can set up a cronjob to restart the server every day.

5) It's a good idea. We already support using the port 587 (instead of 25) but it'd be indeed useful to also support smtp authentication.

6) Unfortunately I don't know much about LDAP. SimpleLogin does support identity providers that support Oauth2 so I wonder if you can maybe modify them to support LDAP. Their code is in https://github.com/simple-login/app/blob/372466ab067f964ca7bf87d50b1e34b60ef8a58f/app/auth/views/facebook.py#L1-L0 for Facebook, https://github.com/simple-login/app/blob/372466ab067f964ca7bf87d50b1e34b60ef8a58f/app/auth/views/github.py#L1-L0 for Github
